src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,216 @@
+
+    private void addToTheEnd(T value) {
+        add(value);
+    }
+
+    private void addToTheBeginning(T value) {
+        Node<T> newNode = new Node<>(null, value, first);
+        first.prev = newNode;
+        first = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item: list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
+        checkIndexValidity(index);
+        if (index < size / 2) {
+            int i = 0;
+            for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+                if (i == index) {
+                    return x.item;
+                }
+                i++;
+            }
+        } else {
+            int i = size - 1;
+            for (MyLinkedList.Node<T> x = last; x != null; x = x.prev) {
+                if (i == index) {
+                    return x.item;
+                }
+                i--;
+            }
+        }
         return null;
     }
 
+    private void checkIndexValidity(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
     @Override
     public T set(T value, int index) {
+        checkIndexValidity(index);
+        if (index < size / 2) {
+            int i = 0;
+            for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+                if (i == index) {
+                    T oldValue = x.item;
+                    x.item = value;
+                    return oldValue;
+                }
+                i++;
+            }
+        } else {
+            int i = size - 1;
+            for (MyLinkedList.Node<T> x = last; x != null; x = x.prev) {
+                if (i == index) {
+                    T oldValue = x.item;
+                    x.item = value;
+                    return oldValue;
+                }
+                i--;
+            }
+        }
         return null;
     }
 
     @Override
     public T remove(int index) {
+        checkIndexValidity(index);
+        int i = 0;
+        for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+            if (i == index) {
+                T removedItem = x.item;
+                remove(x.item);
+                return removedItem;
+            }
+            i++;
+        }
         return null;
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+                if (x.item == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+                if (object.equals(x.item)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
+    void unlink(Node<T> x) {
```

Comment:
The `unlink` method should be private because it's an internal method of your class.